torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -561,7 +561,12 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processMessage(
          // a shared_ptr here.
          rpc = (std::shared_ptr<RpcCommandBase>)std::move(rpc),
          messageType = request.type(),
-         id = request.id()]() {
+         id = request.id(),
+         threadLocalState = ThreadLocalState()]() {
```

Comment:
Is there an actual error that you're seeing or what's the use case for this? Currently, if worker0 invokes the profiler and sends an RPC over to 1, then the profiler on 1 will not run. I'm currently working on a change to enable the profiler to run remotely across RPC nodes, and I agree that for that change we will need this.

Reply:
This is the test I will add. It's not working because the flag, profiler enabled, is now a thread local. I need to add more work to have a process global profiling_enabled flag, which overwrites profiler state of all living threads and new threads to be created.

```cpp
def enable_profiler():
    profiler_config = torch.autograd.ProfilerConfig(torch.autograd.ProfilerState.CPU, False)
    torch.autograd._enable_profiler(profiler_config)
 
 
def disable_profiler():
    records = torch.autograd._disable_profiler()
    events = torch.autograd.profiler.EventList(torch.autograd.profiler.parse_cpu_trace(records))
    return events

class RpcTest(RpcAgentTestFixture):
    @dist_init
    def test_restore_profiler_state_in_continuation(self):
        dst_rank = (self.rank + 1) % self.world_size
        dst_worker_name = worker_name(dst_rank)
 
        x = torch.tensor(1)
        y = torch.tensor(2)
 
        rpc.rpc_sync(dst_worker_name, enable_profiler)
        rpc.rpc_sync(dst_worker_name, torch.add, (x, y))
        events = rpc.rpc_sync(dst_worker_name, disable_profiler)  # Does not work, this is not a global swtich.
 
        last_end_time = 0
        top_level_names_iter = iter(['add'])
        for event in events:
            event_name = event.name
            cpu_interval = event.cpu_interval
            if cpu_interval.start > last_end_time:
                top_level_name_expected = next(top_level_names_iter)
                self.assertEqual(event_name, top_level_name_expected)
                last_end_time = cpu_interval.end
 
        # Ensures the number of events matches expection.
        with self.assertRaisesRegex(StopIteration, ""):
            next(top_level_names_iter)
```

Reply:
I agree that this test would not work, specifically these lines:

```
rpc.rpc_sync(dst_worker_name, enable_profiler)
rpc.rpc_sync(dst_worker_name, torch.add, (x, y))
events = rpc.rpc_sync(dst_worker_name, disable_profiler)
```

the reasoning is that enable_profiler and disable_profiler are not necessarily executed by the same server thread. However, for remote profiler, we don't plan to enable the profiler this way - we plan to do it in `request_callback` right before executing the work (in this case torch.add). Although, we would need to still propagate TLS state such as this profiling flag as in this diff. Let's sync up offline about this.

Reply:
agree with @rohan-varma